flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/reader/AbstractRecordReader.java

```diff
@@ -63,6 +63,8 @@ protected AbstractRecordReader(InputGate inputGate, String[] tmpDirectories) {
 	}
 
 	protected boolean getNextRecord(T target) throws IOException, InterruptedException {
+		inputGate.requestPartitions();
```

Comment:
Why do we have to `requestPartitions()` once per every record?

Reply:
This is the only concentrated place work for all the batch cases. Otherwise we have to add this call for all the specific invokable instances derived from `BatchTask` class, even we also need to consider many other cases which bypassed the invokable class and use the wrapped `AbstractRecordReader` directly in unit tests, etc.

I remembered in the early version, the `requestPartitions` was also placed inside `SingleInputGate#getNext`  method. Because of the mailbox requirement, it was migrated into `#setup` afterwards.

Considering the batch case unrelated to mailbox path, so i think it might be accepted to redo it still in the original place only work for batch cases.

Reply:
Can not we keep the previous setup logic if unaligned checkpoints are disabled/not configured (that would include batch?)? And add a `checkState` somewhere, that unaligned checkpoints can not be used without `StreamTask`/in streaming or something like that?

Reply:
agree, it is not a perfect way, but the feasible simple way ATM to not maintain many different code paths, also not sensitive for batch code path.

Reply:
Could you explain why do you think:
> also not sensitive for batch code path.

?